cpp/src/io/text/multibyte_split.cu

```diff
@@ -0,0 +1,394 @@
+  }
+  cudaEventDestroy(event);
+}
+
+void join_pool_to_stream(rmm::cuda_stream_pool& stream_pool, rmm::cuda_stream_view stream)
+{
+  cudaEvent_t event;
+  cudaEventCreate(&event);
+  for (uint32_t i = 0; i < stream_pool.get_pool_size(); i++) {
+    cudaEventRecord(event, stream_pool.get_stream(i));
+    cudaStreamWaitEvent(stream, event, 0);
+  }
+  cudaEventDestroy(event);
+}
+
+cudf::size_type multibyte_split_scan_full_source(cudf::io::text::data_chunk_source const& source,
+                                                 cudf::io::text::detail::trie const& trie,
+                                                 scan_tile_state<multistate>& tile_multistates,
+                                                 scan_tile_state<uint32_t>& tile_offsets,
+                                                 device_span<cudf::size_type> output_buffer,
+                                                 device_span<char> output_char_buffer,
+                                                 rmm::cuda_stream_view stream,
+                                                 rmm::cuda_stream_pool& stream_pool)
+{
+  CUDF_FUNC_RANGE();
+  cudf::size_type chunk_offset = 0;
+
+  // this function interleaves three kernel executions
+
+  multibyte_split_init_kernel<<<TILES_PER_CHUNK, THREADS_PER_TILE, 0, stream.value()>>>(  //
+    -TILES_PER_CHUNK,
+    TILES_PER_CHUNK,
+    tile_multistates,
+    tile_offsets,
+    cudf::io::text::detail::scan_tile_status::oob);
+
+  auto multistate_seed = multistate();
+
+  multistate_seed.enqueue(0, 0);
+
+  tile_multistates.set_seed_async(multistate_seed, stream);
+  tile_offsets.set_seed_async(0, stream);
+
+  fork_stream_to_pool(stream, stream_pool);
+
+  auto reader = source.create_reader();
+
+  for (auto base_tile_idx = 0; true; base_tile_idx += TILES_PER_CHUNK) {
+    auto chunk_stream = stream_pool.get_stream();
+    auto chunk        = reader->get_next_chunk(ITEMS_PER_CHUNK, chunk_stream);
+
+    if (chunk.size() == 0) { break; }
+
+    // reset the next chunk of tile state
+    multibyte_split_init_kernel<<<TILES_PER_CHUNK, THREADS_PER_TILE, 0, chunk_stream>>>(  //
+      base_tile_idx,
+      TILES_PER_CHUNK,
+      tile_multistates,
+      tile_offsets);
+    multibyte_split_kernel<<<TILES_PER_CHUNK, THREADS_PER_TILE, 0, chunk_stream>>>(  //
+      base_tile_idx,
+      TILES_PER_CHUNK,
+      tile_multistates,
+      tile_offsets,
+      trie.view(),
+      chunk_offset,
+      chunk,
+      output_buffer,
+      output_char_buffer);
+
+    chunk_offset += chunk.size();
+  }
+
+  join_pool_to_stream(stream_pool, stream);
+
+  return chunk_offset;
+}
+
+std::unique_ptr<cudf::column> multibyte_split(cudf::io::text::data_chunk_source const& source,
+                                              std::vector<std::string> const& delimiters,
+                                              rmm::cuda_stream_view stream,
+                                              rmm::mr::device_memory_resource* mr)
+{
+  CUDF_FUNC_RANGE();
```

Comment:
it's very useful when looking at nsight systems to see the two functions side by side, so you can zoom in to see the work that occurs after the first pass before the second pass. Also if we add any caching to the chunk reader, it will be easy to tell how much speed up we get in the second pass.